README.md

```diff
@@ -118,7 +118,7 @@ Use `Ambient_context.create_key` to create an opaque key for the value, and expo
 (* lib/foo_deep.ml *)
 module Ctx = Ambient_context
 
-let context_key : string Ctx.key = Ctx.create_key ()
+let header_context_key : string Ctx.key = Ctx.create_key ()
```

Comment:
Hm — I wanted to prioritize the instructions for application-developers in the docs, since (theoretically) library-developers will need to _link_ here, if they implement `ambient-context` — I want to make sure that user-flow ("I'm using this Foo Lib, do not care about this 'ambient context' business at all, just tell me what I need to do so I can get back to the task at hand, thx."); but yeah, that means the logical/wallclock order are both backwards in the README 😓

Reply:
hrm … should prolly use mdx or executable-readme tests or whatever, in any case. hrmph.

Reply:
Oh, so I was imagining that the more common use case would be application developers that have built up legacy code and want an easy way to inject ad hoc functionality into their deeply nested dependencies. From that perspective, having the instructions flow as "this is how you would hijack the library, then also make sure to add the top level stuff later" seemed more appropriate.

In my mind, `ambient context` was pitched as a last resort so I didn't think other library developers would prioritize implementing it when they could just enforce these "thread-local" values as part of the API interface. For example, passing the http header as a parameter to the API.

That said, you probably meant for `ambient context` as a solution for library maintainers trying to make forward compatible upgrades without breaking it's users. I imagine that unless it's absolutely critical functionality, maintainers would more likely opt out from adding it. If on the other hand, the functionality was really necessary to have, then the library ought to release a new version that breaks users to signal them to upgrade. In which case, now the onus is on the application developers to make a conscious decision to use `ambient context` as a solution for compatibility.

Reply:
Ah, I see the confusion. If you have `app -> bar -> foo`, then this tool is _neither_ about "app author" modifying `foo`, _nor_ about "foo author" being able to modify `foo`'s API — it's _about it being unreasonable for the "bar author" to modify `bar`._

That is, yes, the library-developer you're talking about could totally choose to add a public `~debugging_ctx` argument to every one of their public functions (… which I have personal opinions about, that may or may not be shared by the wider OCaml community … 😓) …

… but to expect _every single intermediate dependency_ to _also_ expose, and pass along, `~debugging_ctx`, is a complete non-starter.

(This is the same overall problem as [W3C's Trace Context](https://www.w3.org/TR/trace-context/) is trying to solve at the "distributed systems" level of the stack; whereas we're solving it in this library in the "dependency tree" layer of the stack.)